File: audira_frontend/lib/core/models/moderation_history.dart
import 'package:equatable/equatable.dart';

/// GA01-163: Modelo para historial de moderaciones
class ModerationHistory extends Equatable {
  final int id;
  final int productId;
  final String productType; // SONG o ALBUM
  final String productTitle;
  final int artistId;
  final String? artistName;
  final String? previousStatus;
  final String newStatus;
  final int moderatedBy;
  final String? moderatorName;
  final String? rejectionReason;
  final DateTime moderatedAt;
  final String? notes;

  const ModerationHistory({
    required this.id,
    required this.productId,
    required this.productType,
    required this.productTitle,
    required this.artistId,
    this.artistName,
    this.previousStatus,
    required this.newStatus,
    required this.moderatedBy,
    this.moderatorName,
    this.rejectionReason,
    required this.moderatedAt,
    this.notes,
  });

  factory ModerationHistory.fromJson(Map<String, dynamic> json) {
    return ModerationHistory(
      id: json['id'] as int,
      productId: json['productId'] as int,
      productType: json['productType'] as String,
      productTitle: json['productTitle'] as String,
      artistId: json['artistId'] as int,
      artistName: json['artistName'] as String?,
      previousStatus: json['previousStatus'] as String?,
      newStatus: json['newStatus'] as String,
      moderatedBy: json['moderatedBy'] as int,
      moderatorName: json['moderatorName'] as String?,
      rejectionReason: json['rejectionReason'] as String?,
      moderatedAt: DateTime.parse(json['moderatedAt'] as String),
      notes: json['notes'] as String?,
    );
  }

  String get newStatusDisplay {
    switch (newStatus) {
      case 'PENDING':
        return 'En revisión';
      case 'APPROVED':
        return 'Aprobado';
      case 'REJECTED':
        return 'Rechazado';
      default:
        return newStatus;
    }
  }

  String get previousStatusDisplay {
    if (previousStatus == null) return '-';
    switch (previousStatus) {
      case 'PENDING':
        return 'En revisión';
      case 'APPROVED':
        return 'Aprobado';
      case 'REJECTED':
        return 'Rechazado';
      default:
        return previousStatus!;
    }
  }

  @override
  List<Object?> get props => [
        id,
        productId,
        productType,
        productTitle,
        artistId,
        artistName,
        previousStatus,
        newStatus,
        moderatedBy,
        moderatorName,
        rejectionReason,
        moderatedAt,
        notes,
      ];
}

File: music-catalog-service/src/main/java/io/audira/catalog/model/ModerationHistory.java
package io.audira.catalog.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Entidad de auditoría para el historial de moderación.
 * <p>
 * Esta tabla es <b>inmutable</b> (solo inserciones). Registra cada cambio de estado
 * en el flujo de aprobación de contenido para cumplir con el requisito <b>GA01-163</b>.
 * Permite responder preguntas como: "¿Quién rechazó este álbum y cuándo?".
 * </p>
 */
@Entity
@Table(name = "moderation_history")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ModerationHistory {

    /** Identificador único del evento de auditoría. */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** ID de la entidad moderada. */
    @Column(name = "product_id", nullable = false)
    private Long productId;

    /** Tipo de producto: "SONG" o "ALBUM". */
    @Column(name = "product_type", nullable = false)
    private String productType;

    /**
     * Snapshot del título del producto al momento de la moderación.
     * <p>Útil si el título cambia posteriormente.</p>
     */
    @Column(name = "product_title", nullable = false)
    private String productTitle;

    /** ID del artista propietario. */
    @Column(name = "artist_id", nullable = false)
    private Long artistId;

    /** Snapshot del nombre del artista. */
    @Column(name = "artist_name")
    private String artistName;

    /** Estado anterior (Origen de la transición). */
    @Enumerated(EnumType.STRING)
    @Column(name = "previous_status")
    private ModerationStatus previousStatus;

    /** Nuevo estado asignado (Destino de la transición). */
    @Enumerated(EnumType.STRING)
    @Column(name = "new_status", nullable = false)
    private ModerationStatus newStatus;

    /** ID del administrador que realizó la acción. */
    @Column(name = "moderated_by", nullable = false)
    private Long moderatedBy;

    /** Nombre del administrador (para visualización rápida en logs). */
    @Column(name = "moderator_name")
    private String moderatorName;

    /**
     * Motivo del rechazo.
     * <p>Generalmente poblado solo si {@code newStatus} es {@code REJECTED}.</p>
     */
    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    /** Fecha y hora exacta del evento. */
    @Column(name = "moderated_at", nullable = false)
    private LocalDateTime moderatedAt;

    /** Notas internas para otros administradores. */
    @Column(columnDefinition = "TEXT")
    private String notes;

    @PrePersist
    protected void onCreate() {
        if (this.moderatedAt == null) {
            this.moderatedAt = LocalDateTime.now();
        }
    }
}
